src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +57,32 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+    std::filesystem::path path(key);
+
+    while (path != _cache_dir) {
+        try {
+            // ss::remove_file removes only empty directory
+            co_await ss::remove_file(path.native());
```

Comment:
maybe we can add a `vassert` before the deletion that checks that the deleted file or directory is inside the `_cache_dir` just in  case (even if it works now doesn't mean the bug may not be introduced later)

Reply:
moved `vassert` from `invalidate` and `evicti` functions to this function since it's the only place the actual deletion happens